lib/icinga/clusterevents-check.cpp

```diff
@@ -93,36 +93,137 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
 	ApiListener::Ptr listener = ApiListener::GetInstance();
 
 	if (!listener) {
-		Log(LogCritical, "ApiListener", "No instance available.");
+		Log(LogCritical, "ApiListener") << "No instance available.";
 		return;
 	}
 
-	if (params->Contains("source")) {
-		Log(LogCritical, "ApiListener", "Not implemented.");
-		return;
-	}
+	Defer resetExecuteCommandProcessFinishedHandler ([]() {
+		Checkable::ExecuteCommandProcessFinishedHandler = nullptr;
+	});
 
-	if (!listener->GetAcceptCommands()) {
-		Log(LogWarning, "ApiListener")
-				<< "Ignoring command. '" << listener->GetName() << "' does not accept commands.";
+	if (params->Contains("source")) {
+		String uuid = params->Get("source");
 
 		Host::Ptr host = new Host();
 		Dictionary::Ptr attrs = new Dictionary();
 
 		attrs->Set("__name", params->Get("host"));
 		attrs->Set("type", "Host");
-		attrs->Set("enable_active_checks", false);
+
+		/*
+		 * Override the check timeout if the parent caller provided the value. Compatible with older versions not
+		 * passing this inside the cluster message.
+		 * This happens with host/service command_endpoint agents and the 'check_timeout' attribute being specified.
+		 */
+		if (params->Contains("check_timeout"))
+			attrs->Set("check_timeout", params->Get("check_timeout"));
 
 		Deserialize(host, attrs, false, FAConfig);
 
 		if (params->Contains("service"))
 			host->SetExtension("agent_service_name", params->Get("service"));
 
-		CheckResult::Ptr cr = new CheckResult();
-		cr->SetState(ServiceUnknown);
-		cr->SetOutput("Endpoint '" + Endpoint::GetLocalEndpoint()->GetName() + "' does not accept commands.");
-		Dictionary::Ptr message = MakeCheckResultMessage(host, cr);
-		listener->SyncSendMessage(sourceEndpoint, message);
+		attrs->Set(params->Get("command_type"), params->Get("command"));
+		attrs->Set("command_endpoint", sourceEndpoint->GetName());
+
+		Deserialize(host, attrs, false, FAConfig);
+		Checkable::Ptr checkable = host;
+		ObjectLock oLock (checkable);
+
+		/* Check deadline */
+		double deadline = params->Get("deadline");
+		if (Utility::GetTime() > deadline) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
```

Comment:
... only for `checkable->GetName()`?

Can't you assemble the name by yourself w/o the deserialization overhead above?